Refactor preprocess and url processing functions

ang-zeyu opened the pull request:
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [x] Other, please explain:

Part of #810 

<!--
    If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"
-->


<!--
    Please ensure your pull request is ready:
    - Bug fix PR that is non-trivial **should** add a page or unit test for regression testing.
    - Feature PR **must** add a page to the user guide for demo.
    - Enhancement PR **should** update the user guide.

    Otherwise, prefix your PR title with "[WIP]".
-->

**What is the rationale for this request?**
Another step to modularising parser functions, enhancing code maintainability and readability.

**What changes did you make? (Give an overview)**
- Break down `_preprocess` and move related functions to a separate file, `componentPreprocessor`
- Refactor common ( to `componentPreprocessor` and `parser` ), stateless url processing functions to `utils/urls`
- Enhance in-code documentation for the above along the way
- Several smaller code refactors for the above ( reducing nested ifs, separating long functions when appropriate, etc. )

**Provide some example code that this change will affect:**

<!-- Paste the example code below: -->
```
// directory structure from parser.js
- parser
  - utils
    - index
    - urls
  - preprocessors
    - componentPreprocessor
...
```

**Is there anything you'd like reviewers to focus on?**

1. Directory structure of the refactor

2. Of note is that `componentPreprocessor` and `parser` is still rather coupled. ( `parser` is passed as a parameter to the entry point in `componentPreprocessor` )
  - Ideally, only `parser` should depend on `componentPreprocessor`, and not the other way around at all if possible.
  - I initially tried making `componentPreprocessor` a class which would subsume required properties from the main `parser`. This was not feasible at the moment however, as many other `parser` functions have 'side effects' on `parser`'s instance variables. It would be counterproductive to add code to synchronize the variables between the two.
  - Hence, we can proceed with a smaller refactor first, and gradually aim to remove this coupling if possible.


**Testing instructions:**
`npm run test` should pass

**Proposed commit message: (wrap lines at 72 characters)**
Refactor preprocess and url processing functions

The parser class houses a large number of functions.
This can be daunting for newer developers to the project, and decreases
maintainability of the code.

Let’s start by modularising the preprocess and url processing functions
into separate files.
Let’s also enhance some in-code documentation for these functions.
Some of these functions can also be abstracted into smaller units of
functionality, increasing the maintainability of the code.

<!--
    See this link for more info on how to write a good commit message:
    https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message
-->

<!--
    Some of the responses that you gave to the previous questions might
    provide you with the information needed to craft your commit message.
-->

ang-zeyu commented:
Ready for review

Some quirks I noticed which I'm not so sure about:

1. Boilerplate include sources are added to the parser's `boilerplateIncludeSrc` array. However, it is also added to `dynamicIncludeSrc` afterward, and is functionally the same for the purposes of live reloading. Should we remove `boilerplateIncludeSrc`? ( left as is for now )

2. `_.hasIn` is used quite a few times to check whether some attribute exists in an element. I've changed these to `_.has`, since attributes are direct properties of `element.attribs`, we do not want to check for indirect attributes.

3. Whether preprocessing logic for thumbnails ( moved to `_preprocessThumbnail` ) should be moved to `componentParser` instead ( left as is for now )

4. For `include` tags, there is some specific processing for `context.mode`. However, the `_preprocess` was never called with anything other than `context.mode = 'include'`. Should we remove this? ( left as is for now )

5. We only call `_rebaseReferenceForStaticIncludes` for `<include>`'s ( left as is )
  5.1. which's includeSrc is an md file,
  5.2. a `src` that has some hash `#...`.
  Is this intentional? ( though, changing it include `5.2` or both `5.2` and `5.1` and running `updatetest` didn't change any output test files )

ang-zeyu commented on src/lib/markbind/src/preprocessors/componentPreprocessor.js:
```diff
@@ -0,0 +1,394 @@
+  console.warn(`<body> tag found in ${node.attribs[ATTRIB_CWF]}. This may cause formatting errors.`);
+}
+
+function preProcessComponent(node, context, config, parser) {
```
Entry point to the original `_preprocess` function, the respective switch cases are the respective control flows of `_preprocess` for different types of elements

ang-zeyu commented on src/lib/markbind/src/preprocessors/componentPreprocessor.js:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    // TODO clarify why this is only done if there is a hash, or fix it if it is unintended
+    if (hash) _rebaseReferenceForStaticIncludes(actualContent, element, config);
```
Is this intentional? ( point 5 )

ang-zeyu commented on src/lib/markbind/src/parser.js:
```diff
@@ -197,193 +178,19 @@ class Parser {
-        element.attribs.boilerplate = element.attribs.boilerplate || path.basename(filePath);
-        actualFilePath
-          = Parser.calculateBoilerplateFilePath(element.attribs.boilerplate, filePath, config);
-        this.boilerplateIncludeSrc.push({ from: context.cwf, to: actualFilePath });
```
Is `boilerplateIncludeSrc` necessary?
For both panels and includes we eventually still push either `dynamicIncludeSrc` or `staticIncludeSrc`

ang-zeyu commented on src/lib/markbind/src/preprocessors/componentPreprocessor.js:
```diff
@@ -0,0 +1,394 @@
+}
+
+// TODO move this to componentParser
+function _preProcessThumbnail(node) {
```
Should this be moved?

yamgent replied:
Whether it should be moved or not will depend on how specialised componentProcessor will eventually get. I noted that there's some coupling between this and the parser that still needs to be resolved in future PRs, so I believe the answer will naturally come by as the refactoring process goes.

Also noted about the weird quirks that you have found. Refactoring by itself is usually quite a tough task by itself; verifying additional changes on top of it is going to make it even more tough, so always welcome to settle those problems in separate PRs (and good to open these in the issue tracker too).

Since my time is now quite restricted as compared to the past, I am no longer able to manually verify that everything works. So these green ticks approval that I give from now on would basically be just a glance through your code to see that there's no major source of issues in terms of design and implementation. You guys might want to test out each other's code from time to time to see if there's any problems.

With all that said, will be merging this. :+1:

ang-zeyu commented on src/lib/markbind/src/preprocessors/componentPreprocessor.js:
```diff
@@ -0,0 +1,394 @@
+      actualContent = isInline ? actualContent : `\n\n${actualContent}\n`;
+    } else {
+      // TODO do we need this branch?
+      actualContent = md.render(actualContent);
```
A quick usage find for `_preprocess` reveals that it is never called with anything other than `context.mode = 'include'`

ang-zeyu commented on src/lib/markbind/src/parser.js:
```diff
@@ -197,193 +178,19 @@ class Parser {
-      if (element.children && element.children.length > 0) {
-        element.children = element.children.map(e => self._preprocess(e, context, config));
-      }
+  extractInnerVariablesIfNotProcessed(content, childContext, config, filePathToExtract) {
```
To allow `componentPreprocessor` and `_extractInnerVariables` to share this common code

ang-zeyu commented on src/lib/markbind/src/parser.js:
```diff
@@ -685,15 +429,71 @@ class Parser {
-      this.preprocess(actualFilePath, pageData, context, config)
-        .then(resolve)
-        .catch(reject);
+      const currentContext = context;
```
This chunk is moved from the `preprocess` ( not `_preprocess` ) function, which was added in the expressive layouts pr,
to remove naming ambiguity.

A future PR may look into whether expressive layouts can be introduced in similar fashion to the page `header`, `sitenav`, etc., as `includeData` ( the function this code is under ) is very similar to `includeFile`. ( to reduce code duplication )

ang-zeyu commented on src/lib/markbind/src/parser.js:
```diff
@@ -804,34 +604,32 @@ class Parser {
-      let newBase = Parser.calculateNewBaseUrls(filePath, this.rootPath, this.baseUrlMap);
-      if (newBase) {
-        const { relative, parent } = newBase;
+      let newBaseUrl = urlUtils.calculateNewBaseUrls(filePath, this.rootPath, this.baseUrlMap);
```
The only irrelevant refactor to remove the nested `ifs`, and slightly better variable naming; Let me know if this should be removed!

acjh commented:
> changing it ... and running `updatetest` didn't change any output test files

Instead of relying on non-exhaustive functional tests, for extensive refactoring, let's check that there is no diff to the generated `_site` of [nus-cs2103/website-base](https://github.com/nus-cs2103/website-base).

ang-zeyu commented:
> Instead of relying on non-exhaustive functional tests, for extensive refactoring, let's check that there is no diff to the generated `_site` of [nus-cs2103/website-base](https://github.com/nus-cs2103/website-base).

Thanks for the suggestion! Will take note of it for upcoming refactors.

I've removed the specific processing for `context.mode`, after confirming the output did not change between the latest master and my branch for the 2103 site. I added a `console.log` in the branch previously as well, and confirmed no statements were printed.

Interestingly, `_rebaseReferenceForStaticIncludes` with any combination of the conditions ( including removing it completely ) did not change the output.
I've left it in for now, it may be better to investigate this in more detail in a seperate PR.

ang-zeyu commented:
Rebased on latest, no changes

yamgent commented:
> Should this be moved?

Whether it should be moved or not will depend on how specialised componentProcessor will eventually get. I noted that there's some coupling between this and the parser that still needs to be resolved in future PRs, so I believe the answer will naturally come by as the refactoring process goes.

Also noted about the weird quirks that you have found. Refactoring by itself is usually quite a tough task by itself; verifying additional changes on top of it is going to make it even more tough, so always welcome to settle those problems in separate PRs (and good to open these in the issue tracker too).

Since my time is now quite restricted as compared to the past, I am no longer able to manually verify that everything works. So these green ticks approval that I give from now on would basically be just a glance through your code to see that there's no major source of issues in terms of design and implementation. You guys might want to test out each other's code from time to time to see if there's any problems.

With all that said, will be merging this. :+1:
